src/backend/joanie/core/models/products.py

```diff
@@ -296,6 +297,81 @@ def generate_certificate(self):
         )
         return certificate
 
+    def proceed_to_payment(self, **kwargs):
+        """Pay order with credit card data or with a registered card.
+        Allow to save credit card for future payments.
+
+        Args:
+            kwargs: dict, contains credit card data
+                name: str, credit card name chosen by user
+                card_number: str, credit card number e.g. 1111222233334444
+                cryptogram: str, credit card cryptogram e.g. 222
+                expiration_date: str, credit card expiration date e.g. '09/21' ('%M%y')
+                save: bool, to register credit card (to a future oneclick payment)
+              or
+                id: uid, credit card uid to get credit card for oneclick payment
+        """
+        paid = False
+        if self.price and self.state != enums.ORDER_STATE_PAID:
+            payment_backend = backends.get_backend()
+            credit_card_uid = kwargs.get("id")
```

Comment:
what I don't like about this approach is that this should have been validated in the API.
The creation of the credit card must be done prior to calling this function on the basis of a call from the frontend directly with the token and last numbers instead of the secret credentials.
If the API does this job, the function `proceed_to_payment` could have a signature:
```python
def proceed_to_payment(self, card_uid, should_save=True):
    ...
```